lib/DefaultFields.js

```diff
@@ -0,0 +1,227 @@
+'use strict'; // eslint-disable-line strict
+
+const Utils = require('./Utils.js');
+const objectCopy = Utils.objectCopy;
+
+/**
+ * This class manages the internal `default fields` for logging classes, with a
+ * notion of parent/child relationships and hierarchy.
+ * A Child element will inherit the fields from the Parent element, and
+ * complete (or even override part of it) with its own fields. For instance,
+ * a child can redefine (and thus override in the final result) a field that is
+ * already defined by its parent.
+ *
+ * This class shall be used embedded within loggers, to provide the `Default
+ * fields` and hierarchy logics.
+ *
+ * @private @property {DefaultFields} parent     - The parent node in the
+ *                                                 hierarchy
+ * @private @property {DefaultFields[]} children - The list of children nodes
+ *                                                 in the hierarchy
+ * @private @property {Object} fields            - The dictionary of fields
+ *                                                 defined for this node
+ * @private @property {Object} parentFields      - The dictionary of fields
+ *                                                 inherited by this node from
+ *                                                 its parent node
+ * @private @property {Object} precomputedFields - The prepared dictionary of
+ *                                                 aggregated fields containing
+ *                                                 the parent's fields
+ *                                                 overriden with the node's
+ *                                                 own fields. This is used as
+ *                                                 a trick to reduce the
+ *                                                 computational need.
+ *
+ * @mixin
+ */
+class DefaultFields {
+    constructor() {
+        this.parent = null;
+        this.children = [];
+        this.fields = {};
+        this.parentFields = {};
+        this.precomputedFields = {};
+    }
+
+    /**
+     * Add a node to the list of children of the current node, effectively
+     * making it one of the node's children.
+     *
+     * @param {DefaultFields} child - The node to add as a child of the current
+     *                                node
+     *
+     * @return {undefined}
+     */
+    _dfRegisterChild(child) {
+        this.children.push(child);
+        this._dfNotifyChild(child);
+    }
+
+    /**
+     * Remove a node from the list of children of the current node, effectively
+     * cutting off the relationship between the two.
+     *
+     * @param {DefaultFields} toRemove - The node to remove from the list of
+     *                                   children of the current node.
+     *
+     * @return {undefined}
+     */
+    _dfUnregisterChild(toRemove) {
+        this.children = this.children.filter(child => child !== toRemove);
```

Comment:
This action might take times. If it could be optimised, it can improve perf. Particularly, it is called from `setParent` that might occur frequently :)